### PR TITLE
New legacy tagging interface

### PR DESCRIPTION
**Pull request by peteglondon**
This PR implements 2 cards
1. [Create new legacy tagging interfaces for Policies, Policy Areas and Specialist Topics out of the edit page in Whitehall](https://trello.com/c/bBTUraJ0/54-create-new-legacy-tagging-interfaces-for-policies-policy-areas-and-specialist-topics-out-of-the-edit-page-in-whitehall)
1. ["New document" flow changes](https://trello.com/c/E8DRoXS2/67-new-document-flow-changes)

It changes the flow for things that can be tagged to the new taxonomy as follows:

Edition Edit Screen > Press 'Next' > Taxonomy Tag screen > Press 'Save topic changes' > Admin page.

The flow for things that do not support the taxonomy is:

Edition Edit Screen > Press 'Next' >Legacy Tagging Screen > Press 'Save legacy associations' > Admin page.

The new legacy tagging page looks like this:
screen shot 2018-05-24 at 14 35 50

### CODE REVIEW

File: features/step_definitions/news_article_steps.rb [hunk 1]
```diff
@@ -117,6 +117,8 @@
   select "", from: "edition_lead_organisation_ids_1"
 
   click_button "Save"
+  # TODO  investigate why save legacy breaks this
```
Comment (cbaines) [inline features/step_definitions/news_article_steps.rb]: Is this still something to do?

File: features/support/admin_legacy_associations_helper.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  def stub_topics
+    @pol_area_1 = create(:topic, name: policy_area_1['title'])
+    @pol_area_3 = create(:topic, name: policy_area_3['title'])
+  end
+
+private
+
+  def assert_selected_policies_are_displayed
+    assert has_css? ".policies li", text: policy_1['title']
+    assert has_css? ".policies li", text: policy_area_1['title'] # TODO this seems wrong
```
Comment (cbaines) [inline features/support/admin_legacy_associations_helper.rb]: I can see that the test does `tag_to_policy_area(policy_area_3)`, I guess that's why the TODO is here?

File: test/functional/edition_legacy_associations_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      title: 'the edition',
+      policy_content_ids: ['5d37821b-7631-11e4-a3cb-005056011aef'],
+      topic_ids: [@topic.id.to_s],
+      primary_specialist_sector_tag: 'aaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaa',
+      secondary_specialist_sector_tags: ['aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeee', 'eeeeeeee-bbbb-cccc-dddd-aaaaaaaaaaaaa']
+    )
+    get :edit, params: { edition_id: @edition.id }
+    assert_select "#edition_policy_content_ids option[value='5d37821b-7631-11e4-a3cb-005056011aef'][selected='selected']"
+    assert_select "#edition_topic_ids option[value='#{@topic.id}'][selected='selected']"
+    # assert_select "#edition_primary_specialist_sector_tag option[value='aaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaa'][selected='selected']"
```
Comment (cbaines) [inline test/functional/edition_legacy_associations_controller_test.rb]: Is it worth uncommenting this, or perhaps removing it?

File: features/step_definitions/news_article_steps.rb (hunk 1 above)
Comment (peteglondon) [reply features/step_definitions/news_article_steps.rb]: Yes it is, it was days ago and I forgot about it. The functionality works when you try it manually, something in the test setup. I'll put it in the card

File: features/support/admin_legacy_associations_helper.rb (hunk 2 above)
Comment (peteglondon) [reply features/support/admin_legacy_associations_helper.rb]: Again I think this is just the test setup, It does the right thing in reality.

File: test/functional/edition_legacy_associations_controller_test.rb (hunk 3 above)
Comment (peteglondon) [reply test/functional/edition_legacy_associations_controller_test.rb]: Actually I forgot to finish the test here, it maybe becomes redundant now as it's covered elsewhere. I'll add it to the card.

File: app/views/admin/edition_legacy_associations/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        html: { class: edition_form_classes(@edition) } do |form| %>
+      <fieldset>
+        <legend>Associations</legend>
+
+        <div class="important-associations">
+          <% if @edition.can_be_related_to_policies? %>
+            <%= render 'admin/shared/policy_fields', form: form %>
+          <% end %>
+          <% if @edition.can_be_associated_with_topics? %>
+            <%= render 'topic_fields', form: form, edition: @edition %>
```
Comment (cbaines) [inline app/views/admin/edition_legacy_associations/edit.html.erb]: Could we take the opportunity to rename this from "topic" to "policy_area", as that would make it a little clearer. The codebase is already inconsistent, so no need to rename everything.

File: app/views/admin/base/_specialist_sector_fields.html.erb [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<fieldset class="edition-specialist-sector-fields">
+ <fieldset class="edition-specialist-sector-fields">
```
Comment (vanitabarrett) [inline app/views/admin/base/_specialist_sector_fields.html.erb]: Some stray space has been added at the start of this line.

File: test/functional/edition_legacy_associations_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    @edition = create(:publication)
+    get :edit, params: { edition_id: @edition.id, return: 'edit'}
+    assert_select ".form-actions a:contains('cancel')[href='#{edit_admin_edition_path(@edition)}']"
+  end
+
+  test 'should update the edition with the selected legacy tags' do
+    @topic = create(:topic)
+    @edition = create(:publication, title: 'the edition')
+
+
```
Comment (vanitabarrett) [inline test/functional/edition_legacy_associations_controller_test.rb]: Stray new line

File: app/views/admin/editions/_show_metadata.html.erb [hunk 7]
```diff
@@ -51,6 +51,12 @@
     selected_taxon_paths: @edition_taxons.map(&:full_path),
     no_topics_message: 'No topics - please add a topic before publishing'
   } %>
+<% else %>
+  <%= render partial: '/admin/shared/tagging/legacy_associations_box', locals: {
+      path_to_edit_associations: edit_admin_edition_legacy_associations_path(@edition.id),
+      edition: @edition,
+      no_policy_areas_message: 'No policy areas - please add a policy area before publishing'
```
Comment (vanitabarrett) [inline app/views/admin/editions/_show_metadata.html.erb]: The `no_policy_areas_message` is only displayed when the edition can be tagged to policy areas. This means that for doc types which can't (I found case studies as an example), no message is shown:

screen shot 2018-05-24 at 11 33 49

Not sure if this is intended or if we want to add a message like "No legacy associations", similar to what we do for translations (as shown).

File: app/views/admin/shared/tagging/_legacy_associations_box.html.erb [hunk 8]
```diff
@@ -0,0 +1,46 @@
+<section class="legacy-associations" id="legacy-associations">
```
Comment (vanitabarrett) [inline app/views/admin/shared/tagging/_legacy_associations_box.html.erb]: The class and id aren't being used, we could probably remove them (or at least the id)

File: lib/linkable_topics.rb [hunk 9]
```diff
@@ -17,6 +21,8 @@ def taxons
 
 private
 
+
```
Comment (vanitabarrett) [inline lib/linkable_topics.rb]: 2 stray new lines here.

File: test/support/admin_edition_controller_test_helpers.rb [hunk 10]
```diff
@@ -71,13 +71,37 @@ def should_allow_creating_of(edition_type)
...
+          edition: attributes
+        }
+
+        edition = edition_class.last
+
+        assert_redirected_to edit_admin_edition_tags_path(edition.id)
+        assert_equal 'The document has been saved', flash[:notice]
+      end
+
+      test "create should take the writer to the legacy tagging page" do
```
Comment (vanitabarrett) [inline test/support/admin_edition_controller_test_helpers.rb]: May be useful to expand the test name to include something like `if can't tag to taxonomy`

File: app/controllers/admin/edition_legacy_associations_controller.rb [hunk 11]
```diff
@@ -3,7 +3,13 @@ class Admin::EditionLegacyAssociationsController < Admin::BaseController
   before_action :enforce_permissions!
   before_action :limit_edition_access!
 
-  def edit; end
+  def edit;
```
Comment (vanitabarrett) [inline app/controllers/admin/edition_legacy_associations_controller.rb]: I don't think the `;` should be here?

File: app/views/admin/editions/_show_metadata.html.erb (hunk 7 above)
Comment (peteglondon) [reply app/views/admin/editions/_show_metadata.html.erb]: I've created a new card to deal with this one. I don't want to hold off getting this one merged as it's holding us all back.

**Pull request merged**